addons/sale_purchase/models/sale_order_line.py

```diff
@@ -159,99 +190,81 @@ def _purchase_service_prepare_line_values(self, purchase_order, quantity=False):
-        product_ctx = {
-            'lang': get_lang(self.env, purchase_order.partner_id.lang).code,
-            'company_id': purchase_order.company_id,
-        }
-        if supplierinfo:
-            price_unit = self.env['account.tax'].sudo()._fix_tax_included_price_company(
-                supplierinfo.price, supplier_taxes, taxes, self.company_id)
-            if purchase_order.currency_id and supplierinfo.currency_id != purchase_order.currency_id:
-                price_unit = supplierinfo.currency_id._convert(price_unit, purchase_order.currency_id, purchase_order.company_id, fields.Date.context_today(self))
-            product_ctx.update({'seller_id': supplierinfo.id})
-        else:
-            product_ctx.update({'partner_id': purchase_order.partner_id.id})
-
-        product = self.product_id.with_context(**product_ctx)
-        name = product.display_name
-        if product.description_purchase:
-            name += '\n' + product.description_purchase
+        price_unit, taxes = self._purchase_service_get_price_unit_and_taxes(supplierinfo, purchase_order)
+        name = self._purchase_service_get_product_name(supplierinfo, purchase_order, quantity)
 
         return {
             'name': name,
             'product_qty': purchase_qty_uom,
             'product_id': self.product_id.id,
             'product_uom': self.product_id.uom_po_id.id,
             'price_unit': price_unit,
-            'date_planned': fields.Date.from_string(purchase_order.date_order) + relativedelta(days=int(supplierinfo.delay)),
+            'date_planned': purchase_order.date_order + relativedelta(days=int(supplierinfo.delay)),
             'taxes_id': [(6, 0, taxes.ids)],
             'order_id': purchase_order.id,
             'sale_line_id': self.id,
         }
 
+    def _purchase_service_match_supplier(self, warning=True):
+        # determine vendor of the order (take the first matching company and product)
+        suppliers = self.product_id._select_seller(quantity=self.product_uom_qty, uom_id=self.product_uom)
+        if warning and not suppliers:
+            raise UserError(_("There is no vendor associated to the product %s. Please define a vendor for this product.") % (self.product_id.display_name,))
+        return suppliers
+
+    def _purchase_service_match_purchase_order(self, partner, company=False):
+        return self.env['purchase.order'].search([
+            ('partner_id', '=', partner.id),
+            ('state', '=', 'draft'),
+            ('company_id', '=', (company and company or self.env.company).id),
+        ], order='id desc')
+
+    def _create_purchase_order(self, supplierinfo):
+        values = self._purchase_service_prepare_order_values(supplierinfo)
+        return self.env['purchase.order'].with_context(mail_create_nosubscribe=True).create(values)
+
+    def _match_or_create_purchase_order(self, supplierinfo):
+        purchase_order = self._purchase_service_match_purchase_order(supplierinfo.partner_id)[:1]
```

Comment:
If we restore the `limit=1` in `_purchase_service_match_purchase_order`, there is no need to limit the output to the first item as the length should already be 0 or 1.
```suggestion
        purchase_order = self._purchase_service_match_purchase_order(supplierinfo.partner_id)
```

Reply:
moved to second commit